lib: add throws option to fs.f/l/statSync

amcasey opened the pull request:
For consumers that aren't interested in *why* a `statSync` call failed,
allocating and throwing an exception is an unnecessary expense.  This PR
adds an option that will cause it to return `undefined` in such cases
instead.

As a motivating example, the JavaScript & TypeScript language service
shared between Visual Studio and Visual Studio Code is stuck with
synchronous file IO for architectural and backward-compatibility
reasons.  It frequently needs to speculatively check for the existence
of files and directories that may not exist (and cares about file vs
directory, so `existsSync` is insufficient), but ignores file system
entries it can't access, regardless of the reason.

Benchmarking the language service is difficult because it's so hard to
get good coverage of both code bases and user behaviors, but, as a
representative metric, we measured batch compilation of a few hundred
popular projects (by star count) from GitHub and found that, on average,
we saved about 1-2% of total compilation time.  We speculate that the
savings could be even more significant in interactive (language service
or watch mode) scenarios, where the same (non-existent) files need to be
polled over and over again.  It's not a huge improvement, but it's a
very small change and it will affect a lot of users (and CI runs).

For reference, our measurements were against `v12.x` (3637a061a at the
time) on an Ubuntu Server desktop with an SSD.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

BridgeAR commented on doc/api/fs.md:
```diff
@@ -2166,6 +2166,9 @@ changes:
 * `options` {Object}
   * `bigint` {boolean} Whether the numeric values in the returned
     [`fs.Stats`][] object should be `bigint`. **Default:** `false`.
+  * `noThrow` {boolean} Whether an exception should be thrown
```
Using a negation is not ideal for options. It's more difficult than just e.g., `throws` and using `true` as default.

amcasey replied:
My concern was that people already passing `{ bigInt: true }` would need to update their code.

amcasey replied:
Happy to choose a new name, I just think `undefined` has to retain the current behavior.

amcasey replied:
How about `returnUndefinedOnError`?

lundibundi replied:
> My concern was that people already passing { bigint: true } would need to update their code.

I think @BridgeAR meant to have check like `if (options.throws === false)` so that if the user hasn't specified `false` explicitly then it would work as it is now so there won't be any breaking. (this retains the current behavior with `undefined`)

IMO `throws` is clearer for option name.

amcasey replied:
@lundibundi If that's idiomatic, I'm happy to change it, but I have to admit that I find it confusing that the falsey values `null` and `undefined` would mean `true`.

lundibundi replied:
I understand your concern but it's not uncommon to have such cases. The API clearly states that `boolean` is expected so if non-boolean is passed (when nothing is passed `undefined` is just interpreted as default === true) it is free to just throw a `TypeError` (for i.e `null`).

Would the `ignoreErrors` option name with `false` default sound better perhaps?

amcasey replied:
I always try to adopt local conventions.  I people are accustomed to `undefined` meaning "use the default", then `throw` is a great name.

amcasey replied:
Done.  Thanks for clarifying!

BridgeAR commented:
@amcasey I guess you already use the `withFileTypes` option while using `readdir()`?

amcasey commented:
@BridgeAR Yes, discovering that was a huge win - thanks!

amcasey commented:
> There is another way to improve performance in case you expect many errors that you are not interested in: using `Error.stackTraceLimit`

I tried this on the same data set I used to validate my change and saw a change that was close enough to zero that I'd want to do a lot more iterations before declaring one way or the other.  (For simplicity, I just set `Error.stackTraceLimit = 0` at the start of the program and left it - let me know if that somehow invalidated the experiment.)

BridgeAR commented:
@amcasey it would depend if any other code resets the stack trace limit to a different value.

I just ran a quick benchmark in the REPL:

```console
> console.time(); var limit = Error.stackTraceLimit; Error.stackTraceLimit = 0; for (let i = 0; i < 1e6; i++) try { throw new Error() } catch {}; Error.stackTraceLimit = limit; console.timeEnd()
default: 2.380s
> console.time(); for (let i = 0; i < 1e6; i++) try { throw new Error() } catch {}; console.timeEnd()
default: 7.260s
```

amcasey commented:
@BridgeAR We only have one explicit access to the property and it's not hit in the case I tested.  Is it possible that some other API we're calling reset it?  Also, I didn't mean to imply that it doesn't work in any scenario, just that it doesn't obviously work in ours.

BridgeAR commented:
> Is it possible that some other API we're calling reset it?

That is absolutely possible. You could check what is set while exiting (e.g., `process.on('exit', () => console.log(Error.stackTraceLimit))`).

lundibundi commented on lib/fs.js:
```diff
@@ -1013,29 +1013,47 @@ function stat(path, options = { bigint: false }, callback) {
   const ctx = { fd };
   const stats = binding.fstat(fd, options.bigint, undefined, ctx);
-  handleErrorFromBinding(ctx);
+  if (options.throw === false) {
```
Sorry for misunderstanding but could you use `throws` instead of `throw` to avoid conflicting with keyword?

amcasey replied:
My bad. 😅

lundibundi commented on test/parallel/test-fs-stat-bigint.js:
```diff
@@ -122,6 +122,23 @@ if (!common.isWindows) {
+}
+
+{
+  runSyncFailureTest(fs.fstatSync, 9999, { code: 'EBADF' });
```
Nit: I'd just inline these to make the clearer/simpler.

amcasey replied:
Personally, I find this way more readable.  However, I'll switch it if anyone else supports the change.

lundibundi commented:
ping @nodejs/fs @BridgeAR

MicahZoltu commented:
Beyond just the operational overhead cost, debugging errors can become more difficult when code is throwing exceptions during normal operations.  If you debug an application and set it to break on all exceptions, it will break every time some piece of code tries to check for file existence.  For something like the TypeScript compiler, this means hundreds or thousands of *expected* exceptions as the compiler probes for files on disk (file not found is a very much expected result).  While in some cases you can work around this by setting a breakpoint after the file probing is complete and then enabling uncaught exceptions, this strategy doesn't work when the exception you are trying to break on happens in the middle of a large amount of file probing, or when the application you are debugging just naturally does a lot of file probing.

lundibundi commented:
ping @BridgeAR @nodejs/fs

bnoordhuis commented:
Let's start with the bits I agree with. :-)

Yes, `fs.statSync()` is expensive when most calls throw. It's why I changed `require()` to use something else a long time ago.

However, rather than adding a `noThrow` option, why not use `fs.accessSync()` instead? It should be faster on Windows - and probably other platforms too - because it does less work (fewer system calls, less data conversion.)

MicahZoltu commented:
Looking at the docs, that doesn't appear to address the problem of throwing being expensive, since that function still throws in all failure scenarios:
> If any of the accessibility checks fail, an Error will be thrown. Otherwise, the method will return undefined.

amcasey commented:
`accessSync` also doesn't indicate file system entry type (etc), so you'd end up calling `statSync` immediately afterward.

lundibundi commented:
ping @BridgeAR @bnoordhuis @nodejs/fs

joyeecheung commented on lib/fs.js:
```diff
@@ -1013,29 +1013,47 @@ function stat(path, options = { bigint: false }, callback) {
   const stats = binding.fstat(fd, options.bigint, undefined, ctx);
-  handleErrorFromBinding(ctx);
+  if (options.throws === false) {
+    if (ctx.errno || ctx.error) {
```
This would return whatever left inside the stats buffer (probably from the last call) if there's no error which seems to be a bizarre way to leak the stats buffer to the users. Can you just return undefined here `if (options.throws === false)`? (Also the same below)

joyeecheung replied:
Actually I think it should be

```js
if (ctx.errno || ctx.error) {
  if (options.throws === false) {
    return undefined;
  }
}
handleErrorFromBinding(ctx);
return getStatsFromBinding(stats);
```

amcasey replied:
@joyeecheung I'm not sure I follow.  Which path through the ifs are you concerned about?  If I'm reading your suggested update correctly, it looks like it only differs in calling `handleErrorFromBinding` in the case where no exception is desired and no error occurred (which I expected to be a no-op).

Also, leaking stale stats seems very bad, but I'm not sure I understand how that could occur.  Can you please elaborate?

joyeecheung replied:
Ah, sorry, I looked into it again and I think you are right, when there's no error and `(options.throws === false)` `handleErrorFromBinding(ctx)` can be skipped.

amcasey replied:
Thanks for checking!  Better safe than sorry. :smile:

joyeecheung commented:
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/663/

amcircleci commented:
@joyeecheung Can you please point me at the failure in `node-test-commit-windows-fanned`?  I found this but it looks non-fatal:

```
[ RUN      ] EnvironmentTest.EnvironmentWithNoESMLoader
(node:5404) ExperimentalWarning: VM Modules is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
(node:5404) UnhandledPromiseRejectionWarning: Error: Not supported
    at vm:module(0):1:1
    at SourceTextModule.evaluate (internal/vm/module.js:224:23)
    at embedder_main_12.js:1:328
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(node:5404) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:5404) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
[       OK ] EnvironmentTest.EnvironmentWithNoESMLoader (79 ms)
```

The other stack traces in the file look like Jenkins infrastructure issues (possibly as a result of something upstream I haven't identified).

amcasey commented:
@joyeecheung Can you please point me at the failure in `node-test-commit-windows-fanned`?  I found this but it looks non-fatal:

```
[ RUN      ] EnvironmentTest.EnvironmentWithNoESMLoader
(node:5404) ExperimentalWarning: VM Modules is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
(node:5404) UnhandledPromiseRejectionWarning: Error: Not supported
    at vm:module(0):1:1
    at SourceTextModule.evaluate (internal/vm/module.js:224:23)
    at embedder_main_12.js:1:328
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(node:5404) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:5404) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
[       OK ] EnvironmentTest.EnvironmentWithNoESMLoader (79 ms)
```

The other stack traces in the file look like Jenkins infrastructure issues (possibly as a result of something upstream I haven't identified).

amcasey commented:
I'm going to close and reopen in the hope that that retriggers CI.

lundibundi commented:
@amcasey no need to worry, it is entirely possible for CI to fail without being affected by your changes. The most recent CI looks like internal failures to me.

> Is there some way to confirm that it actually did have this change?

You can see in the log the patch https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/663/console 
```cmd
23:19:37 ++ mv out/Release/node ./node-master
23:19:37 ++ case $USE_CASE in
23:19:37 ++ curl -L https://github.com/nodejs/node/pull/33716.patch
23:19:37 ++ git apply -3
```

The problem here is that in the benchmark configuration there are both options `['throw', 'noThrow']` which will result in *both* of them being run on the current master, which obviously doesn't support the `throws: false` option yet and will therefore always fail.
Also, I don't think there is a way to run `compare` benchmark for this. I'm not sure what would be the best way here since this benchmark will not be correctly compared (since it needs to compare `between-options` instead of `between-node-versions-for-different-options`) between node versions which is the main use-case of these benchmarks (see https://github.com/nodejs/node/blob/master/doc/guides/writing-and-running-benchmarks.md#comparing-nodejs-versions)

amcasey commented:
@lundibundi Thanks for investigating!  I could add a try-catch to the `noThrow` case as well.  It would, presumably, show no improvement in master and a slightly reduced improvement with this change (assuming try-catch has overhead, as in some other languages).  Happy to do that if it would increase confidence in this PR.

lundibundi commented:
@amcasey I don't think that is needed, it will work just fine once this lands. The issue here is with how we benchmark itself, it's not tailored to "compare a vs b" but instead to "compare a to a on different node versions".
IMO the benchmark is fine as is and since the performance difference you showed here https://github.com/nodejs/node/pull/33716#issuecomment-686012760 is quite drastic. I'd say it is an improvement.

Let's wait a bit to allow @bnoordhuis to perhaps add something here since he expressed his concerns with this API. If everything is okay I expect this to land soon. Big thanks for this contribution :+1:

JasonKleban commented:
Bump

amcasey commented:
Force pushed a rebase. No substantive changes (or even manual merges).

MylesBorins commented:
Kicking off CI again. if it is green I'll land this tomorrow

MylesBorins commented:
Still seeing failures in CI but they do not seem to be related to this change. I rebased and am kicking off CI one more time

MylesBorins commented:
So it looks like we are seeing `parallel/test-http2-respond-file-error-pipe-offset` as the primary source of flakiness here, which seems unrelated to this change

https://github.com/nodejs/node/issues/35881

tniessen reviewed (dismissed):
Just my two cents, and sorry for chiming in super late: This seems like a dangerous optimization to me. Sure, performance is important, but shouldn't correctness be the primary concern?

For example, [the `stat` function](https://man7.org/linux/man-pages/man2/stat.2.html) may fail with the error code `ENOMEM` if the kernel runs out of memory, and that is completely unrelated to the existence and availability of the file. Are we just going to pretend that the file is inaccessible without providing any error information whatsoever? That doesn't sound like a solid plan to me.

On Windows, `stat` requires more than one system call. As far as I can tell, at least [`CreateFileW`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew) and [`NtQueryInformationFile`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntqueryinformationfile) are called, and the function fails if either system call fails. Now, I don't know all possible error conditions of these system calls, but I bet some of them are, again, completely unrelated to the existence/availability of the file itself.

Designing an API to allow consumers to blindly ignore errors regardless of potentially unexpected causes does not seem like a safe approach to me, and I value safety and correctness more than a 1% performance gain.

tniessen commented:
Possible middle ground: Only ignore specific error codes (e.g., `ENOENT`, `ENOTDIR`, etc.), but that probably depends on the system calls on each supported platform.

tniessen commented:
> Scoping to a specific set of error codes would be just fine, though it will take a little thought to design an API that adequately conveys that.

It's just an idea to avoid ignoring unexpected error causes, I haven't really thought it through. I was thinking, something along the lines of a boolean option `throwIfNoEntry`.

Maybe someone else has a better idea for solving this problem.

> How would you feel about passing in a list of ignored error codes?

That sounds like a very generic approach. If we add this to one API, people will probably start asking for the same feature on all other synchronous APIs. I don't know if the small performance gains (in error cases) justify the added complexity, and aggregating the list is not trivial either, I assume.

---

> It frequently needs to speculatively check for the existence of files and directories that may not exist (and cares about file vs directory, so `existsSync` is insufficient),

We don't usually optimize for error code paths. If exceptions really are the performance bottleneck in your application, did you try something like this to avoid exceptions?

```js
try {
  if (existsSync(path)) {
    if (statSync(path).isFile()) {
      // Cool, the file exists! Now do something with it.
    }
  }
} catch (err) {
  // Super rare error condition! The file exists, but we weren't able to stat it.
}
```

The performance impact depends on the ratio of existing to non-existing files, I guess. If most files exist and are files, this adds `existsSync`, but in that case, you'd also have very few exceptions, so I assume quite a lot of your file lookups are failing, in which case this version avoids exceptions.

amcasey commented:
> I was thinking, something along the lines of a boolean option throwIfNoEntry.
I've switched to `throwIfNoEntry`.  Please let me know if I've misunderstood.

> We don't usually optimize for error code paths.
For our use case, these aren't errors - some files exist and some files don't.  When you're doing module resolution, you generally look at several non-existent paths before you find an existing module.

> did you try something like [pre-checking `existsSync`] to avoid exceptions?
Yes, but there isn't a strong, consistent imbalance between existent and non-existent files, so we didn't come out ahead.  In places where we don't need the actual `Stats`, we switched to `existsSync` and it helped.

amcasey commented:
I didn't redo my product perf testing, but the node benchmarks indicate not throwing is still ~7X as fast as throwing, so I don't think the extra logic for checking the error code made much difference.

tniessen commented:
Thank you for taking my concerns into consideration @amcasey. I will withdraw my request for changes.

tniessen commented:
PTAL @nodejs/fs @joyeecheung @lundibundi

targos commented on test/parallel/test-fs-stat-bigint.js:
```diff
@@ -122,6 +122,28 @@ if (!common.isWindows) {
+}
+
+{
+  assert.throws(() => fs.fstatSync(9999), { code: 'EBADF' });
```
```suggestion
  assert.throws(() => fs.fstatSync(9999), { code: 'EBADF' });
  assert.throws(() => fs.fstatSync(9999, { throwIfNoEntry: false }), { code: 'EBADF' });
```

amcasey replied:
I left that off because it's not a declared part of the API, but I see no harm in testing it.

amcasey replied:
Oops, that caused a lint error.  Fix pending.

amcasey commented:
Is this just a timeout?
```
collect2: fatal error: ld terminated with signal 9 [Killed]
```

amcasey commented:
Can the failures be ignored as unrelated?  If not, what's next?  I think I'll need help for either a re-run or a diagnosis.  Thanks!

MylesBorins commented:
@amcasey we do need to get a green CI run. If the same test continues to fail and it is unrelated we should mark it flaky and get that approved separately before landing

amcasey commented:
Thanks, @MylesBorins!  I'm happy to work within the process - I just didn't know what it was.

amcasey commented:
In `node-test-linux-linked-debug`, I see the same apparent-timeout as before:
> collect2: fatal error: ld terminated with signal 9 [Killed]

In `node-test-linux-linked-zlib` I see what appears to be a test failure in `sequential/test-cpu-prof-worker-argv`.  I'm not familiar with that test, so I'm investigating.

amcasey commented:
Force pushed a rebase fixing a merge conflict (another change also added an import to fs.js).

amcasey commented:
Looks like a different timeout this time?
> Build timed out (after 120 minutes). Marking the build as failed.

amcasey commented:
@MylesBorins What's the next step for getting a green CI run?  Could my JS-only change have slowed down the native build enough to cause a timeout?  If so, is there something I can do to reproduce the failure locally so I can investigate?  Thanks!

MylesBorins commented:
Trying to run CI one more time... we may want to rebase again. I'm not 100% why this would be failing here... but it is consistently failing and we are not seeing those failures in other prs

amcasey commented:
@MylesBorins I'm perfectly willing to believe it's my change - I just don't know how to investigate.  Is there some way I can kick off CI runs myself?  Then I can selectively revert parts of the change until I find the problem.

I'm just heading into a meeting, but I should be able to push a rebase within the next two hours, if that works for you.

amcasey commented:
@MylesBorins - the rebased branch is passing local tests.  Do you want me to push or do you want to stick with the green run we have now?

MylesBorins commented:
@amcasey the green run we have now is good enough. landing

MylesBorins commented:
Landed in 3772cf56ae28

amcasey commented:
Thanks @MylesBorins!

merceyz commented on doc/api/fs.md:
```diff
@@ -2568,6 +2568,9 @@ changes:
 * `options` {Object}
   * `bigint` {boolean} Whether the numeric values in the returned
     [`fs.Stats`][] object should be `bigint`. **Default:** `false`.
+  * `throwIfNoEntry` {boolean} Whether an exception will be thrown
+    if no file system entry exists, rather than returning `undefined`.
+    **Default:** `true`.
```
Could the history be updated to reflect when this was added please?

amcasey replied:
@merceyz Sorry, I'm not familiar with that process.  Is there a changelog or something that needs to be updated?  Can you point me at a sample or a doc?

merceyz replied:
I assume the history a few lines above this is where it's added but I'm also not familiar with it

amcasey replied:
Huh, I didn't see that.  I guess I'm hoping it's generated, because I don't know how I'd know what version to use.

merceyz replied:
It isn't i'm afraid cc @MylesBorins

amcasey replied:
Some of the history items above (e.g. https://github.com/nodejs/node/commit/7b5b8ef42474c7370740c1feefc41330c44ce1ea) make it look like the changelog was updated?

merceyz replied:
It's in the changelog yes but not the history https://nodejs.org/docs/latest-v15.x/api/fs.html#fs_fs_statsync_path_options

amcasey replied:
I see.  So you'd like a new PR with `Added in: v15.3.0` in each of the new sections?

merceyz replied:
Yes please :)
